feature: support SPI scope

CharmingRabbit opened the pull request:
### Ⅰ. Describe what this PR did
Optimize SPI
1.Provide unified caching for singleton extension
2..Support to distinguish singleton and multiple when load extension
3..Adjust the original code to load the extension using the new API

### Ⅱ. Does this pull request fix one issue?
#2346
### Ⅲ. Why don't you add test cases (unit test/integration test)?
### Ⅳ. Describe how to verify it

### Ⅴ. Special notes for reviews
Note that when loading an extension now, the activateName is case sensitive.
TODO:
The caller's redundant code  which used to load  and cache the extension themselves will be  optimized later.

zjinlei commented on server/src/main/java/io/seata/server/lock/db/DataBaseLocker.java:
```diff
@@ -33,7 +34,7 @@
  * @author zhangsen
  */
-@LoadLevel(name = "db")
+@LoadLevel(name = "db",scope = Scope.PROTOTYPE)
```
SINGLETON

zjinlei commented on server/src/main/java/io/seata/server/session/DefaultSessionManager.java:
```diff
@@ -31,7 +32,7 @@
  * @author sharajava
  */
-@LoadLevel(name = "default")
+@LoadLevel(name = "default",scope = Scope.PROTOTYPE)
```
SINGLETON

zjinlei replied:
my mistake

zjinlei commented on server/src/main/java/io/seata/server/session/db/DataBaseSessionManager.java:
```diff
@@ -46,7 +47,7 @@
  * @author zhangsen
  */
-@LoadLevel(name = "db")
+@LoadLevel(name = "db",scope = Scope.PROTOTYPE)
```
SINGLETON

zjinlei commented on server/src/main/java/io/seata/server/session/file/FileBasedSessionManager.java:
```diff
@@ -45,7 +46,7 @@
  * @author slievrly
  */
-@LoadLevel(name = "file")
+@LoadLevel(name = "file",scope = Scope.PROTOTYPE)
```
SINGLETON

zjinlei commented on server/src/main/java/io/seata/server/store/db/DatabaseTransactionStoreManager.java:
```diff
@@ -54,7 +55,7 @@
  * @author zhangsen
  */
-@LoadLevel(name = "db")
+@LoadLevel(name = "db",scope = Scope.PROTOTYPE)
```
SINGLETON

zjinlei commented on server/src/main/java/io/seata/server/store/file/FileTransactionStoreManager.java:
```diff
@@ -56,7 +57,7 @@
  * @author slievrly
  */
-@LoadLevel(name = "file")
+@LoadLevel(name = "file",scope = Scope.PROTOTYPE)
```
SINGLETON

slievrly commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -336,11 +422,51 @@ private static void loadFile(Class<?> service, String dir, ClassLoader classLoad
+        //do cache
+        extensionClassUrlMap.put(clazz, result);
+        if (loadLevel != null) {
+            if (extensionNameUrlsMap.containsKey(serviceName)) {
```
different SPI loadLevel‘s name can be same.

slievrly commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -336,11 +422,51 @@ private static void loadFile(Class<?> service, String dir, ClassLoader classLoad
+        if (priority >= highestLoadPriority) {
+            highestLoadPriority = priority;
+            defaultExtensionURL = result;
+            return defaultExtensionURL;
```
remove

CharmingRabbit replied:
The highest priority based on the order param in LoadLevel matches the current class's  defaultExtensionURL,we cached them together so that next time this extension is loaded,we can directly find the default extension by cache.

slievrly commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -179,133 +199,197 @@
-    private static <S> List<Class> findAllExtensionClass(Class<S> service, String activateName, ClassLoader loader) {
-        List<Class> extensions = new ArrayList<>();
+    private Map<ExtensionURL, Class<?>> findAllExtensionClass(ClassLoader loader) {
+        Map<ExtensionURL, Class<?>> extensions = new HashMap<>();
```
why not declare as TreeMap？

CharmingRabbit replied:
yes ,treeMap is more appropriate.It has been changed

CharmingRabbit replied:
Cause that treeMap uses comparable() to judge whether the extensionURLs are the same, but we need to judge by the hashCode() of extensionURLs, so here we can only declare as HashMap

zjinlei commented:
please resolve conflicts

zjinlei commented:
![image](https://user-images.githubusercontent.com/20024955/75523204-84c5fd80-5a46-11ea-9282-b961b430a4a4.png)

zjinlei commented on server/src/main/java/io/seata/server/session/db/DataBaseSessionManager.java:
```diff
@@ -79,7 +79,8 @@ public DataBaseSessionManager(String name) {
 
```
DataBaseSessionManager's taskName is different, so its Scope is PROTOTYPE.

CharmingRabbit commented:
The list of Extended implementation to be confirmed
1.Package:discovery Extension Point:RegistryProvider
(1)ConsulRegistryProvider 
(2)CustomRegistryProvider
(3)CustomRegistryProviderForTest
(4)ZookeeperRegistryProvider 
(5)EtcdRegistryProvider 
(6)SofaRegistryProvider 
(7)EurekaRegistryProvider
(8)RedisRegistryProvider 
(9)NacosRegistryProvider 

2.Package:discovery Extension Point:AbstractLoadBalance
(1)RoundRobinLoadBalance 
(2)RandomLoadBalance 

3.Package:metric Extension Point:Exporter
(1)PrometheusExporter 

4.Package:metrics Extension Point:Registry
(1)CompactRegistry 

5.Package:config Extension Point:ConfigurationProvider
(1)EtcdConfigurationProvider
(2)ConsulConfigurationProvider 
(3)SpringCloudConfigurationProvider 
(4)ZookeeperConfigurationProvider
(5)ApolloConfigurationProvider 
(6)NacosConfigurationProvider 
(7)CustomConfigurationProvider 
(8)CustomConfigurationProviderForTest 

6.Package:test Extension Point:LoaderTestSPI
(1)LoaderTestImpl1
(2)LoaderTestImpl2 

7.Package:common(test) Extension Point:Hello
(1)FrenchHello 
(2)EnglishHello 
(3)ChineseHello 
(4)LatinHello

8.Package:rm.datasource Extension Point:AbstractTableMetaCache
(1)OracleTableMetaCache 
(2)MysqlTableMetaCache 
(3)PostgresqlTableMetaCache

9.Package:rm.datasource Extension Point:AbstractUndoLogManager
(1)MySQLUndoLogManager 
(2)OracleUndoLogManager   
(2)PostgresqlUndoLogManager

10.Package:server Extension Point:AbstractTransactionStoreManager
(1)DatabaseTransactionStoreManager

11.Package:server Extension Point:SessionManager
(1)DataBaseSessionManager (Prototype,Confirmed)
(2)DefaultSessionManager (Prototype,Confirmed)
(3)FileBasedSessionManager (Prototype,Confirmed)

12.Package:core Extension Point:LogStore
(1)LogStoreDataBaseDAO

13.Package:core Extension Point:LockStore
(1)LockStoreDataBaseDAO 

14.Package:core Extension Point:Compressor
(1)GzipCompressor
(2)ZipCompressor

15.Package:core Extension Point:ContextCore
(1)ThreadLocalContextCore 
(2)FastThreadLocalContextCore 

16.Package:core Extension Point:AbstractLocker
(1)DataBaseLocker
(2)MemoryLocker (Prototype,Confirmed)

17.Package:core Extension Point:Compressor
(1)NoneCompressor 
(2)SevenZCompressor 
(3)BZip2Compressor 
(4)ZipCompressor

18.Package:core Extension Point:AbstractDataSourceGenerator
(1)DbcpDataSourceGenerator
(2)DruidDataSourceGenerator 

19.Package:core Extension Point:Serializer
(1)HessianSerializer 
(2)KryoSerializer 
(3)ProtobufSerializer  
(4)SeataSerializer 

20.Package:rm.datasource Extension Point:KeywordChecker
(1)MySQLKeywordChecker
(2)OracleKeywordChecker 
(3)PostgresqlKeywordChecker

21.Package:rm.datasource Extension Point:UndoLogParser
(1)JacksonUndoLogParser 
(2)KryoUndoLogParser 
(3)FastjsonUndoLogParser
(4)ProtostuffUndoLogParser

22.Package:rm.datasource Extension Point:SQLRecognizerFactory
(1)DruidDelegatingSQLRecognizerFactory

23.Package:rm.datasource Extension Point:UndoExecutorHolder
(1)MySQLUndoExecutorHolder
(2)OracleUndoExecutorHolder
(3)PostgresqlUndoExecutorHolder

24.Package:sqlparser Extension Point:DbTypeParser
(1)DruidDelegatingDbTypeParser

20.Package:sqlparser Exporter:SQLOperateRecognizerHolder
(1)MySQLOperateRecognizerHolder
(2)OracleOperateRecognizerHolder
(3)PostgresqlOperateRecognizerHolder

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -42,77 +45,100 @@
+     * @param <S> the type
+     * @return the service loader
+     */
+    public static <S> EnhancedServiceLoader<S> getServiceLoader(Class<S> type) {
```
Keep the EnhancedServiceLoader's api unchanged?

CharmingRabbit replied:
OK，I've restore the old api

zjinlei commented:
> The list of Extended implementation to be confirmed:
> 1.Package:discovery Extension Point:RegistryProvider
> (1)ConsulRegistryProvider
> (2)CustomRegistryProvider
> (3)CustomRegistryProviderForTest
> (4)ZookeeperRegistryProvider
> (5)EtcdRegistryProvider
> (6)SofaRegistryProvider
> (7)EurekaRegistryProvider
> (8)RedisRegistryProvider
> (9)NacosRegistryProvider
> 
> 2.Package:discovery Extension Point:AbstractLoadBalance
> (1)RoundRobinLoadBalance
> (2)RandomLoadBalance
> 
> 3.Package:metric Extension Point:Exporter
> (1)PrometheusExporter
> 
> 4.Package:metrics Extension Point:Registry
> (1)CompactRegistry
> 
> 5.Package:config Extension Point:ConfigurationProvider
> (1)EtcdConfigurationProvider
> (2)ConsulConfigurationProvider
> (3)SpringCloudConfigurationProvider
> (4)ZookeeperConfigurationProvider
> (5)ApolloConfigurationProvider
> (6)NacosConfigurationProvider
> (7)CustomConfigurationProvider
> (8)CustomConfigurationProviderForTest
> 
> 6.Package:test Extension Point:LoaderTestSPI
> (1)LoaderTestImpl1
> (2)LoaderTestImpl2
> 
> 7.Package:common(test) Extension Point:Hello
> (1)FrenchHello
> (2)EnglishHello
> (3)ChineseHello
> (4)LatinHello
> 
> 8.Package:rm.datasource Extension Point:AbstractTableMetaCache
> (1)OracleTableMetaCache
> (2)MysqlTableMetaCache
> (3)PostgresqlTableMetaCache
> 
> 9.Package:rm.datasource Extension Point:AbstractUndoLogManager
> (1)MySQLUndoLogManager
> (2)OracleUndoLogManager
> (2)PostgresqlUndoLogManager
> 
> 10.Package:server Extension Point:AbstractTransactionStoreManager
> (1)DatabaseTransactionStoreManager
> 
> 11.Package:server Extension Point:SessionManager
> (1)DataBaseSessionManager(Prototype,Confirmed)
> (2)DefaultSessionManager(Prototype,Confirmed)
> (3)FileBasedSessionManager(Prototype,Confirmed)
> 
> 12.Package:core Extension Point:LogStore
> (1)LogStoreDataBaseDAO
> 
> 13.Package:core Extension Point:LockStore
> (1)LockStoreDataBaseDAO
> 
> 14.Package:core Extension Point:ContextCore
> (1)ThreadLocalContextCore
> (2)FastThreadLocalContextCore
> 
> 15.Package:core Extension Point:AbstractLocker
> (1)DataBaseLocker
> (2)MemoryLocker(Prototype,Confirmed)
> 
> 16.Package:core Extension Point:Compressor
> (1)NoneCompressor
> (2)SevenZCompressor
> (3)BZip2Compressor
> (4)ZipCompressor
> (5)GzipCompressor
> 
> 17.Package:core Extension Point:AbstractDataSourceGenerator
> (1)DbcpDataSourceGenerator
> (2)DruidDataSourceGenerator
> 
> 18.Package:core Extension Point:Serializer
> (1)HessianSerializer
> (2)KryoSerializer
> (3)ProtobufSerializer
> (4)SeataSerializer
> 
> 19.Package:rm.datasource Extension Point:KeywordChecker
> (1)MySQLKeywordChecker
> (2)OracleKeywordChecker
> (3)PostgresqlKeywordChecker
> 
> 20.Package:rm.datasource Extension Point:UndoLogParser
> (1)JacksonUndoLogParser
> (2)KryoUndoLogParser
> (3)FastjsonUndoLogParser
> (4)ProtostuffUndoLogParser
> 
> 21.Package:rm.datasource Extension Point:SQLRecognizerFactory
> (1)DruidDelegatingSQLRecognizerFactory
> 
> 22.Package:rm.datasource Extension Point:UndoExecutorHolder
> (1)MySQLUndoExecutorHolder
> (2)OracleUndoExecutorHolder
> (3)PostgresqlUndoExecutorHolder
> 
> 23.Package:sqlparser Extension Point:DbTypeParser
> (1)DruidDelegatingDbTypeParser
> 
> 24.Package:sqlparser Exporter:SQLOperateRecognizerHolder
> (1)MySQLOperateRecognizerHolder
> (2)OracleOperateRecognizerHolder
> (3)PostgresqlOperateRecognizerHolder

1、2、5、8、9、14、16、17、18、19、20、21、22、23、24 are SINGLETON.
6、7 are used by test only.
@CharmingRabbit @ggndnn Please review.

CharmingRabbit commented:
> > The list of Extended implementation to be confirmed:
> > 1.Package:discovery Extension Point:RegistryProvider
> > (1)ConsulRegistryProvider
> > (2)CustomRegistryProvider
> > (3)CustomRegistryProviderForTest
> > (4)ZookeeperRegistryProvider
> > (5)EtcdRegistryProvider
> > (6)SofaRegistryProvider
> > (7)EurekaRegistryProvider
> > (8)RedisRegistryProvider
> > (9)NacosRegistryProvider
> > 2.Package:discovery Extension Point:AbstractLoadBalance
> > (1)RoundRobinLoadBalance
> > (2)RandomLoadBalance
> > 3.Package:metric Extension Point:Exporter
> > (1)PrometheusExporter
> > 4.Package:metrics Extension Point:Registry
> > (1)CompactRegistry
> > 5.Package:config Extension Point:ConfigurationProvider
> > (1)EtcdConfigurationProvider
> > (2)ConsulConfigurationProvider
> > (3)SpringCloudConfigurationProvider
> > (4)ZookeeperConfigurationProvider
> > (5)ApolloConfigurationProvider
> > (6)NacosConfigurationProvider
> > (7)CustomConfigurationProvider
> > (8)CustomConfigurationProviderForTest
> > 6.Package:test Extension Point:LoaderTestSPI
> > (1)LoaderTestImpl1
> > (2)LoaderTestImpl2
> > 7.Package:common(test) Extension Point:Hello
> > (1)FrenchHello
> > (2)EnglishHello
> > (3)ChineseHello
> > (4)LatinHello
> > 8.Package:rm.datasource Extension Point:AbstractTableMetaCache
> > (1)OracleTableMetaCache
> > (2)MysqlTableMetaCache
> > (3)PostgresqlTableMetaCache
> > 9.Package:rm.datasource Extension Point:AbstractUndoLogManager
> > (1)MySQLUndoLogManager
> > (2)OracleUndoLogManager
> > (2)PostgresqlUndoLogManager
> > 10.Package:server Extension Point:AbstractTransactionStoreManager
> > (1)DatabaseTransactionStoreManager
> > 11.Package:server Extension Point:SessionManager
> > (1)DataBaseSessionManager(Prototype,Confirmed)
> > (2)DefaultSessionManager(Prototype,Confirmed)
> > (3)FileBasedSessionManager(Prototype,Confirmed)
> > 12.Package:core Extension Point:LogStore
> > (1)LogStoreDataBaseDAO
> > 13.Package:core Extension Point:LockStore
> > (1)LockStoreDataBaseDAO
> > 14.Package:core Extension Point:ContextCore
> > (1)ThreadLocalContextCore
> > (2)FastThreadLocalContextCore
> > 15.Package:core Extension Point:AbstractLocker
> > (1)DataBaseLocker
> > (2)MemoryLocker(Prototype,Confirmed)
> > 16.Package:core Extension Point:Compressor
> > (1)NoneCompressor
> > (2)SevenZCompressor
> > (3)BZip2Compressor
> > (4)ZipCompressor
> > (5)GzipCompressor
> > 17.Package:core Extension Point:AbstractDataSourceGenerator
> > (1)DbcpDataSourceGenerator
> > (2)DruidDataSourceGenerator
> > 18.Package:core Extension Point:Serializer
> > (1)HessianSerializer
> > (2)KryoSerializer
> > (3)ProtobufSerializer
> > (4)SeataSerializer
> > 19.Package:rm.datasource Extension Point:KeywordChecker
> > (1)MySQLKeywordChecker
> > (2)OracleKeywordChecker
> > (3)PostgresqlKeywordChecker
> > 20.Package:rm.datasource Extension Point:UndoLogParser
> > (1)JacksonUndoLogParser
> > (2)KryoUndoLogParser
> > (3)FastjsonUndoLogParser
> > (4)ProtostuffUndoLogParser
> > 21.Package:rm.datasource Extension Point:SQLRecognizerFactory
> > (1)DruidDelegatingSQLRecognizerFactory
> > 22.Package:rm.datasource Extension Point:UndoExecutorHolder
> > (1)MySQLUndoExecutorHolder
> > (2)OracleUndoExecutorHolder
> > (3)PostgresqlUndoExecutorHolder
> > 23.Package:sqlparser Extension Point:DbTypeParser
> > (1)DruidDelegatingDbTypeParser
> > 24.Package:sqlparser Exporter:SQLOperateRecognizerHolder
> > (1)MySQLOperateRecognizerHolder
> > (2)OracleOperateRecognizerHolder
> > (3)PostgresqlOperateRecognizerHolder
> 
> 1、2、5、8、9、14、16、17、18、19、20、21、22、23、24 are SINGLETON.
> 6、7 are used by test only.
> @CharmingRabbit @ggndnn Please review.

I've updated the list and adjusted the scope property value in the code

CharmingRabbit commented:
为方便大家review，更快地了解改动、设计意图，在此罗列出对EnhancedServiceLoader做出的主要改动点：
1.核心点：每个扩展，都会对应一个唯一的ServiceURL。
2.一个扩展点的所有扩展实现在加载完毕后，会将加载优先级最高的作为本扩展点的默认实现进行缓存，当不传递name获取时，将获取此默认实现。
3.调用方获取扩展时，无论是否根据name获取，都能获得唯一对应的一个ServiceURL。
4.根据name获取扩展：一个name可能有多个扩展实现（用extensionNameURLsMap缓存），加载完成后会按order排好序并缓存，可保证获取到优先级最高的实现。
5.不传递name获取扩展，会返回默认的优先级最高的实现。
6.缓存：所有扩展的class都会进行缓存，并会缓存class与serviceURL的映射（extensionClassURLMap）；单例扩展的实例会进行缓存，多例扩展实例不缓存，每次创建最新。
7.单例多例，通过LoadLevel的Scope指定，单例加载后将从统一缓存中获取，多例每次回返回新实例。

ggndnn commented on common/src/main/java/io/seata/common/loader/ExtensionURL.java:
```diff
@@ -0,0 +1,101 @@
+ *
+ * @author: haozhibei
+ */
+public class ExtensionURL {
```
The class name is easy to misunderstand. What about ExtensionDefinition or ExtensionInfo.

CharmingRabbit replied:
access left default done and how about ExtensionIdentifier?

ggndnn replied:
I think this class is not used to distinguish extensions, but describes extensions, may be include more attributes in the future, so ExtensionIdentifier (extension ID) is not very suitable. In my opinion, definition is more suitable.

CharmingRabbit replied:
yeah,i agree with ServiceDefinition

CharmingRabbit replied:
Yeah,i agree with ExtensionDefinition

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +177,463 @@
+                    synchronized (holder) {
+                        instance = holder.get();
+                        if (instance == null) {
+                            instance = createExtension(url, loader, argTypes, args);
```
Why not use createNewExtension here directly?

CharmingRabbit replied:
Our purpose is to distinguish whether the extension point is Singleton or multiple. When the LoadLevel's prototype was set to singleton, we return the same instance (using cache), while it was set to Prototype, we use CreateNewExtension() to create new instance

ggndnn replied:
In the 415 row, use createNewExtension instead of createExtension.

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +177,463 @@
-            if (e instanceof EnhancedServiceNotFoundException) {
-                throw (EnhancedServiceNotFoundException)e;
+            if (Scope.SINGLETON.equals(url.getScope())) {
+                Holder<Object> holder = extensionInstances.get(url);
```
What's the benefit to use Holder here? Is Holder class necessary?

CharmingRabbit replied:
This part of the code which use Holder and Double-Checkd Locking refers to Dubbo SPI. I think it has two functions at present:
1. When concurrent, ensure that an extensionDefinition use the same lock object
2. Whether the value in the Holder is null can be used as the judgment condition of Double-Checked Locking
Without using Holder here,i think  it is difficult to guarantee the above two points at the same time.

CharmingRabbit replied:
This part of the code which use Holder and Double-Checkd Locking refers to Dubbo SPI. I think it has two functions at present:
1.When concurrent, ensure that an extensionDefinition use the same lock object
2.Whether the value in the Holder is null can be used as the judgment condition of Double-Checked Locking
Without using Holder here,i think it is difficult to guarantee the above two points at the same time.

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +177,463 @@
-            if (a1 != null) {
-                o1 = a1.order();
+        private S createNewExtension(ExtensionURL url, ClassLoader loader, Class[] argTypes, Object[] args) {
+            Class<?> clazz = loadAllExtensionClass(loader).get(url);
```
If ExtensionURL has the service-class as field, calling loadAllExtensionClass is not necessary.

CharmingRabbit replied:
This part of the code which use Holder and Double-Checkd Locking refers to Dubbo SPI. I think it has two functions at present:
1.When concurrent, ensure that an extensionDefinition use the same lock object
2.Whether the value in the Holder is null can be used as the judgment condition of Double-Checked Locking
Without using Holder here,i think it is difficult to guarantee the above two points at the same time.

CharmingRabbit replied:
Yeah,good idea,i add serviceClass field and change the data structure of  Map（ExtensionDefinition_To_Class） to List<ExtensionDefintion>

ggndnn commented on common/src/main/java/io/seata/common/util/Holder.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Helper Class for hold a value.
+ *
+ * @author: haozhibei
```
@author: xxx -> @author xxx

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +177,463 @@
-                        foundFromCache = false;
-                        providers.put(service, extensions);
-                    }
+    static class InnerEnhancedServiceLoader<S> {
```
The access modifiers here should be private.

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +177,463 @@
+         * @param <S>  the type
+         * @return the service loader
+         */
+        static <S> InnerEnhancedServiceLoader<S> getServiceLoader(Class<S> type) {
```
The access modifiers here should be private.

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +177,463 @@
+         * @return s s
+         * @throws EnhancedServiceNotFoundException the enhanced service not found exception
+         */
+        public S load(ClassLoader loader) throws EnhancedServiceNotFoundException {
```
The access modifiers here should be private.

CharmingRabbit replied:
changed

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +177,463 @@
+                    extensionNameUrlsMap.put(serviceName, urls);
+                }
+            }
+            if (priority >= highestLoadPriority) {
```
Since priority defaults to 0, is the statement better to be `if (defaultExtensionURL == null || priority > highestLoadPriority)`?

CharmingRabbit replied:
the highestLoadPriority is set to Integer.MIN_VALUE,so even if the default priority is 0, the old statement still works,but your statement will be easier to understand here.

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +177,463 @@
+                //defaultExtensionURL,so cached them together
+                highestLoadPriority = priority;
+                defaultExtensionURL = result;
+                return defaultExtensionURL;
```
The `return` statement can be removed.

CharmingRabbit replied:
removed

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +177,463 @@
+            ExtensionURL result = new ExtensionURL(serviceName, typeName, priority, scope);
+            //do cache
+            extensionClassUrlMap.put(clazz, result);
+            if (loadLevel != null) {
```
Should the `if` statement be `if (serviceName != null)`, or other more suitable one?

CharmingRabbit replied:
Since if loadLevel is not null,the name will be required and not null,we can use either loadLevel or serviceName !=null here,to make it easier to understand, changed to (serviceName != null)

ggndnn commented on common/src/main/java/io/seata/common/loader/ExtensionDefinition.java:
```diff
@@ -0,0 +1,111 @@
+ */
+class ExtensionDefinition {
+    private String name;
+    private String typeName;
```
Since we have serviceClass, typeName is a little redundant.

CharmingRabbit replied:
Ok,remove typeName

ggndnn commented on core/src/main/java/io/seata/core/store/StoreMode.java:
```diff
@@ -25,26 +25,35 @@
-     * @param mode the mode
-     * @return the store mode
+     * get value of store mode
+      * @param name the mode name
```
Format correctly.

CharmingRabbit replied:
OK

ggndnn commented on core/src/main/java/io/seata/core/serializer/SerializerFactory.java:
```diff
@@ -72,4 +72,4 @@ public static Serializer getSerializer(byte serializeCode) {
     }
 
-}
+}
```
This kind modification is not necessary.

CharmingRabbit replied:
Got it

ggndnn commented on metrics/seata-metrics-core/src/main/java/io/seata/metrics/exporter/ExporterType.java:
```diff
@@ -26,7 +26,17 @@
      * Export metrics data to Prometheus
      */
-    PROMETHEUS;
+    PROMETHEUS("prometheus");
```
Just a question about this kind modification. Why not ignore case in EnhancedServiceLoader#load, keeping the original enum implementation unchanged?

CharmingRabbit replied:
First, name will be used as a judgment point to distinguish whether two extensiondefinitions are the same，more over I want users to use our spi in a more rigorous and standard way

ggndnn commented on rm-datasource/src/main/java/io/seata/rm/datasource/undo/KeywordCheckerFactory.java:
```diff
@@ -39,8 +39,9 @@ public static KeywordChecker getKeywordChecker(String dbType) {
         }
-        KeywordChecker tableMetaCache = EnhancedServiceLoader.load(KeywordChecker.class, dbType);
+        KeywordChecker tableMetaCache =
+                EnhancedServiceLoader.load(KeywordChecker.class, dbType);
```
I think there is no uniform format style yet, so it's better to keep it unchanged.

CharmingRabbit replied:
Restored

ggndnn commented on server/src/main/java/io/seata/server/session/SessionHolder.java:
```diff
@@ -89,15 +89,19 @@ public static void init(String mode) throws IOException {
-            ROOT_SESSION_MANAGER = EnhancedServiceLoader.load(SessionManager.class, StoreMode.DB.name());
-            ASYNC_COMMITTING_SESSION_MANAGER = EnhancedServiceLoader.load(SessionManager.class, StoreMode.DB.name(),
+            ROOT_SESSION_MANAGER =
+                    EnhancedServiceLoader.load(SessionManager.class, StoreMode.DB.getName());
```
Try to keep the format as original as possible and reduce modification.

CharmingRabbit replied:
OK

ggndnn commented on server/src/test/java/io/seata/server/coordinator/DefaultCoordinatorMetricsTest.java:
```diff
@@ -67,7 +67,7 @@ public void test() throws IOException, TransactionException, InterruptedExceptio
 
             //we need sleep for a short while because default canBeCommittedAsync() is true
-            Thread.sleep(200);
+            Thread.sleep(1000);
```
Is 200 not enough?

CharmingRabbit replied:
Restored

ggndnn commented on common/src/main/java/io/seata/common/loader/ExtensionDefinition.java:
```diff
@@ -0,0 +1,101 @@
+import io.seata.common.util.StringUtils;
+
+/**
+ * The type URL
```
Wrong class name.

CharmingRabbit replied:
Corrected

ggndnn commented on common/src/main/java/io/seata/common/loader/ExtensionDefinition.java:
```diff
@@ -0,0 +1,101 @@
+        if (obj == null) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) {
```
Sub class is not allowed?

CharmingRabbit replied:
I think the class ExtensionDefinition should not be inherited. How about setting it to final?

ggndnn replied:
fine.

CharmingRabbit replied:
issue was added [2454](https://github.com/seata/seata/issues/2454)

ggndnn commented on common/src/main/java/io/seata/common/loader/ExtensionDefinition.java:
```diff
@@ -0,0 +1,101 @@
+        if (!scope.equals(other.scope)) {
+            return false;
+        }
+        return true;
```
Can be optimized to `return !scope.equals(other.scope`.

CharmingRabbit replied:
Ok,done

ggndnn commented on common/src/test/java/io/seata/common/loader/EnhancedServiceLoaderTest.java:
```diff
@@ -98,4 +98,32 @@ public void getAllExtensionClass1() {
+                assertThat(!hellows2.contains(hello));
+            }
+        }
+    }
```
Wrong usage of `assertThat`, should be `assertThat(xxxx).isTrue()` or `assertThat(xxxx).isFalse()`.

CharmingRabbit replied:
Changed to the right usage

ggndnn commented on common/src/test/java/io/seata/common/loader/EnhancedServiceLoaderTest.java:
```diff
@@ -98,4 +98,32 @@ public void getAllExtensionClass1() {
+    public void getMultipleExtensionInstance(){
+        Hello hello1 = EnhancedServiceLoader.load(Hello.class, "LatinHello");
+        Hello hello2 = EnhancedServiceLoader.load(Hello.class, "LatinHello");
+        assertThat(hello1 != hello2);
```
LatinHello is SINGLETON not PROTOTYPE.

CharmingRabbit replied:
My fault , add scope = prototype in LatinHello

ggndnn commented on common/src/main/java/io/seata/common/loader/ExtensionDefinition.java:
```diff
@@ -0,0 +1,101 @@
+
+    public void setScope(Scope scope) {
+        this.scope = scope;
+    }
```
I think all setters here are useless, should be removed.

CharmingRabbit replied:
Removed

ggndnn commented on common/src/test/java/io/seata/common/loader/EnhancedServiceLoaderTest.java:
```diff
@@ -34,8 +34,8 @@
     public void testLoadByClassAndClassLoader() {
-        Hello load = EnhancedServiceLoader.load(Hello.class, Hello.class.getClassLoader());
-        Assertions.assertEquals(load.say(), "Bonjour");
+        Hello load = EnhancedServiceLoader.load(Hello.class,Hello.class.getClassLoader());
```
Should keep format unchanged as before, missing space character here.

CharmingRabbit replied:
restored

ggndnn commented on common/src/test/java/io/seata/common/loader/EnhancedServiceLoaderTest.java:
```diff
@@ -54,15 +54,15 @@ public void testLoadException() {
     @Test
     public void testLoadByClassAndActivateName() {
-        Hello englishHello = EnhancedServiceLoader.load(Hello.class, "EnglishHello");
+        Hello englishHello = EnhancedServiceLoader.load(Hello.class,"EnglishHello");
```
Should keep format unchanged as before, missing space character here.

CharmingRabbit replied:
Restored the missing space character.

ggndnn commented on core/src/main/java/io/seata/core/store/StoreMode.java:
```diff
@@ -25,26 +25,35 @@
-     * @return the store mode
+     * get value of store mode
+     * @param name the mode name
+     *  @return the store mode
```
Format issue here.

CharmingRabbit replied:
Formated

ggndnn commented on server/src/main/java/io/seata/server/lock/db/DataBaseLocker.java:
```diff
@@ -50,8 +51,8 @@ public DataBaseLocker() {
-        lockStore = EnhancedServiceLoader.load(LockStore.class, StoreMode.DB.name(), new Class[] {DataSource.class},
-            new Object[] {logStoreDataSource});
+        lockStore = EnhancedServiceLoader.load(LockStore.class, StoreMode.DB.getName(), new Class[] {DataSource.class},
+                new Object[] {logStoreDataSource});
```
Should keep format unchanged as before, redundant tab character here.

CharmingRabbit replied:
Ok,removed the redundant tab character.

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +173,445 @@
+                    nameToDefinitionsMap.put(serviceName, definitions);
+                }
+            }
+            if (defaultExtensionDefinition == null || priority >= highestLoadPriority) {
```
As said before, this statement should be `if (defaultExtensionDefinition == null || priority > highestLoadPriority)`.

CharmingRabbit replied:
This part of code was removed cause that they wer unnecessary.

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +173,445 @@
-    }
+            if (!extensionDefinitions.isEmpty()) {
+                Collections.sort(extensionDefinitions, (definition1, definition2) -> {
+                    int o1 = definition1.getOrder();
```
Since extensionDefinitions were sorted, is the defaultExtensionDefinition calculating in `getExtensionDefinition` 
 necessary?

CharmingRabbit replied:
Yeah, it is true that the calculating in getExtensionDef is  unnecessary,I will remove them.

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +173,445 @@
+                }
             }
+            result = definitions.stream().map(def -> def.getServiceClass()).collect(Collectors.toList());
+            return result;
```
Remove var `result` and `return definitions.stream().map(ExtensionDefinition::getServiceClass).collect(Collectors.toList())`?

CharmingRabbit replied:
OK, done

ggndnn commented:
@CharmingRabbit Please resolve conflicts.

CharmingRabbit commented:
> @CharmingRabbit Please resolve conflicts.

Resolved,please review.

CharmingRabbit commented:
@slievrly @ph3636 
PTAL

slievrly commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -84,7 +77,8 @@
      */
     public static <S> S load(Class<S> service, String activateName) throws EnhancedServiceNotFoundException {
-        return loadFile(service, activateName, findClassLoader());
+        return InnerEnhancedServiceLoader.getServiceLoader(service).load(activateName,
```
It is recommended that findClassLoader() be placed in the EnhancedServiceLoader class, which is responsible for the parameter assembly.

CharmingRabbit replied:
OK,now i place it in the EnhancedServiceLoader, and uniformly construct the parameters before calling the inner method.

slievrly commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +173,446 @@
+        @SuppressWarnings("rawtypes")
+        private S loadExtension(String activateName, ClassLoader loader, Class[] argTypes,
+                                Object[] args) {
+            if (io.seata.common.util.StringUtils.isEmpty(activateName)) {
```
In the original logic, this could be empty.

CharmingRabbit replied:
This method is used for scenarios where the user explicitly passes activateName, not the initial loaded scenario.

ph3636 commented:
DataBaseSessionManager should be PROTOTYPE

CharmingRabbit commented:
> DataBaseSessionManager should be PROTOTYPE

OK, changed it to Prototype

slievrly commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +173,446 @@
+        private final Holder<List<ExtensionDefinition>> definitionsHolder = new Holder<>();
+        private final ConcurrentMap<ExtensionDefinition, Holder<Object>> definitionToInstanceMap =
+                new ConcurrentHashMap<>();
+        private final ConcurrentMap<String, List<ExtensionDefinition>> nameToDefinitionsMap = new ConcurrentHashMap<>();
```
When is this a collection instead of a single element?

CharmingRabbit replied:
I think we can allow an extension implementation to have the same name and interface name, so here a name may correspond to multiple implementations

slievrly commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +173,446 @@
-                                LOGGER.warn("load [{}] class fail. {}", line, e.getMessage());
+
+        @SuppressWarnings("rawtypes")
+        private void loadFile(String dir, ClassLoader classLoader, List<ExtensionDefinition> extensions)
```
how about set？ What happens if I write two identical records in the spi file?

CharmingRabbit replied:
Without adjusting the data structure, now I will check whether the ExtensionDefinition has been loaded . If it is the same two-line configuration, I will print a warn log and skip the repeated loading

slievrly commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +173,446 @@
-        } catch (IOException e) {
-            throw new EnhancedServiceNotFoundException(e);
+        private S createNewExtension(ExtensionDefinition definition, ClassLoader loader, Class[] argTypes, Object[] args) {
+            loadAllExtensionClass(loader);
```
if necessary？

CharmingRabbit replied:
I checked all the call stacks of createNewExtension(), found that when createNew() is executed, all extension classes have been loaded, so it's unnecessary.I will remove it.

CharmingRabbit commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +173,445 @@
+                    } catch (Throwable e) {
+                        LOGGER.warn(e.getMessage());
+                    } finally {
+                        IOUtil.close(reader);
```
Changed

CharmingRabbit commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +173,445 @@
+                    List<ExtensionDefinition> definitions = new ArrayList<>();
+                    definitions.add(result);
+                    nameToDefinitionsMap.put(serviceName, definitions);
+                }
```
Nice simplification, done

ggndnn replied:
I think the origin meaning was to change the code from `line 513 - 517` to `nameToDefinitionsMap.computeIfAbsent(serviceName, e -> new ArrayList<>()).add(result)`.

CharmingRabbit replied:
> I think the origin meaning was to change the code from `line 513 - 517` to `nameToDefinitionsMap.computeIfAbsent(serviceName, e -> new ArrayList<>()).add(result)`.

Yeah，My understanding is wrong, now changed

CharmingRabbit commented on metrics/seata-metrics-exporter-prometheus/src/main/java/io/seata/metrics/exporter/prometheus/PrometheusExporter.java:
```diff
@@ -37,7 +38,7 @@
  * @author zhengyangyong
  */
-@LoadLevel(name = "Prometheus", order = 1)
+@LoadLevel(name = "prometheus", order = 1, scope = Scope.PROTOTYPE)
```
Changed.

CharmingRabbit commented on metrics/seata-metrics-registry-compact/src/main/java/io/seata/metrics/registry/compact/CompactRegistry.java:
```diff
@@ -37,7 +38,7 @@
  * @author zhengyangyong
  */
-@LoadLevel(name = "Compact", order = 1)
+@LoadLevel(name = "compact", order = 1, scope = Scope.PROTOTYPE)
```
Changed

CharmingRabbit commented on serializer/seata-serializer-seata/src/main/java/io/seata/serializer/seata/SeataSerializer.java:
```diff
@@ -28,7 +29,7 @@
  * @author zhangsen
  */
-@LoadLevel(name = "SEATA")
+@LoadLevel(name = "SEATA", scope = Scope.PROTOTYPE)
```
Changed

slievrly commented:
> LockStoreDataBaseDAO

check 13

slievrly commented:
> The list of Extended implementation to be confirmed:
> 1.Package:discovery Extension Point:RegistryProvider （All of the following are singleton）
> (1)ConsulRegistryProvider
> (2)CustomRegistryProvider
> (3)CustomRegistryProviderForTest
> (4)ZookeeperRegistryProvider
> (5)EtcdRegistryProvider
> (6)SofaRegistryProvider
> (7)EurekaRegistryProvider
> (8)RedisRegistryProvider
> (9)NacosRegistryProvider
> 
> 2.Package:discovery Extension Point:AbstractLoadBalance（All of the following are singleton）
> (1)RoundRobinLoadBalance
> (2)RandomLoadBalance
> 
> 3.Package:metric Extension Point:Exporter (All of the following are singleton)
> (1)PrometheusExporter
> 
> 4.Package:metrics Extension Point:Registry (All of the following are singleton)
> (1)CompactRegistry
> 
> 5.Package:config Extension Point:ConfigurationProvider（All of the following are singleton）
> (1)EtcdConfigurationProvider
> (2)ConsulConfigurationProvider
> (3)SpringCloudConfigurationProvider
> (4)ZookeeperConfigurationProvider
> (5)ApolloConfigurationProvider
> (6)NacosConfigurationProvider
> (7)CustomConfigurationProvider
> (8)CustomConfigurationProviderForTest
> 
> 6,7 are just for test use,so deleted
> 
> 8.Package:rm.datasource Extension Point:AbstractTableMetaCache（All of the following are singleton）
> (1)OracleTableMetaCache
> (2)MysqlTableMetaCache
> (3)PostgresqlTableMetaCache
> 
> 9.Package:rm.datasource Extension Point:AbstractUndoLogManager（All of the following are singleton）
> (1)MySQLUndoLogManager
> (2)OracleUndoLogManager
> (2)PostgresqlUndoLogManager
> 
> 10.DatabaseTransactionStoreManager (No longer SPI)
> 
> 11.Package:server Extension Point:SessionManager
> (1)DataBaseSessionManager(Prototype,Confirmed)
> (2)FileBasedSessionManager(Prototype,Confirmed)
> 
> 12.LogStoreDataBaseDAO （No longer SPI）
> 
> 13.LockStoreDataBaseDAO (No longer SPI)
> 
> 14.Package:core Extension Point:ContextCore（All of the following are singleton）
> (1)ThreadLocalContextCore
> (2)FastThreadLocalContextCore
> 
> 15.Package:core Extension Point:AbstractLocker (No longer SPI)
> 
> 16.Package:core Extension Point:Compressor（All of the following are singleton）
> (1)NoneCompressor
> (2)SevenZCompressor
> (3)BZip2Compressor
> (4)ZipCompressor
> (5)GzipCompressor
> 
> 17.Package:core Extension Point:AbstractDataSourceGenerator（All of the following are singleton）
> (1)DbcpDataSourceGenerator
> (2)DruidDataSourceGenerator
> 
> 18.Package:core Extension Point:Serializer（All of the following are singleton）
> (1)HessianSerializer
> (2)KryoSerializer
> (3)ProtobufSerializer
> (4)SeataSerializer
> 
> 19.Package:rm.datasource Extension Point:KeywordChecker（All of the following are singleton）
> (1)MySQLKeywordChecker
> (2)OracleKeywordChecker
> (3)PostgresqlKeywordChecker
> 
> 20.Package:rm.datasource Extension Point:UndoLogParser（All of the following are singleton）
> (1)JacksonUndoLogParser
> (2)KryoUndoLogParser
> (3)FastjsonUndoLogParser
> (4)ProtostuffUndoLogParser
> 
> 21.Package:rm.datasource Extension Point:SQLRecognizerFactory（All of the following are singleton）
> (1)DruidDelegatingSQLRecognizerFactory
> 
> 22.Package:rm.datasource Extension Point:UndoExecutorHolder（All of the following are singleton）
> (1)MySQLUndoExecutorHolder
> (2)OracleUndoExecutorHolder
> (3)PostgresqlUndoExecutorHolder
> 
> 23.Package:sqlparser Extension Point:DbTypeParser（All of the following are singleton）
> (1)DruidDelegatingDbTypeParser
> 
> 24.Package:sqlparser Exporter:SQLOperateRecognizerHolder（All of the following are singleton）
> (1)MySQLOperateRecognizerHolder
> (2)OracleOperateRecognizerHolder
> (3)PostgresqlOperateRecognizerHolder

lack  `LockManager`、`AbstractCore`

slievrly commented:
![image](https://user-images.githubusercontent.com/8758457/77312701-774a1d80-6d3d-11ea-9bf0-3fa0bb7d2b11.png)

ggndnn commented on common/src/test/java/io/seata/common/loader/EnhancedServiceLoaderTest.java:
```diff
@@ -98,4 +98,32 @@ public void getAllExtensionClass1() {
+    public void getSingletonExtensionInstance(){
+        Hello hello1 = EnhancedServiceLoader.load(Hello.class, "ChineseHello");
+        Hello hello2 = EnhancedServiceLoader.load(Hello.class, "ChineseHello");
+        assertThat(hello1 == hello2);
```
Expression `assertThat` needs `.isTrue()`.

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -113,15 +106,8 @@
      * @throws EnhancedServiceNotFoundException the enhanced service not found exception
      */
     public static <S> S load(Class<S> service, String activateName, Object[] args)
```
If service is SINGLETON and we call `load` twice with different args, what instance should return? So far we get the first created instance.

CharmingRabbit replied:
We can mark as an issue and reslove it later.

ggndnn replied:
fine.

ggndnn commented on common/src/test/java/io/seata/common/loader/EnhancedServiceLoaderTest.java:
```diff
@@ -98,4 +98,32 @@ public void getAllExtensionClass1() {
+        List<Hello> hellows2 = EnhancedServiceLoader.loadAll(Hello.class);
+        for (Hello hello : hellows1){
+            if(hello.say()!="Olá."){
+                assertThat(hellows2.contains(hello));
```
Expression `assertThat` needs `.isTrue()`.

CharmingRabbit replied:
Added

ggndnn commented on common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java:
```diff
@@ -198,181 +172,417 @@
+        private ExtensionDefinition getDefaultExtensionDefinition() {
+            List<ExtensionDefinition> currentDefinitions = definitionsHolder.get();
+            if (currentDefinitions != null && currentDefinitions.size() > 0) {
+                return definitionsHolder.get().get(currentDefinitions.size() - 1);
```
definitionsHolder.get() -> currentDefinitions

CharmingRabbit replied:
done
